pkg/storegateway/indexheader/thread.go

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/thread.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"runtime"
+)
+
+type execResult struct {
+	value interface{}
+	err   error
+	// populated is only set to `true` by the start method of OSThread. The field only
+	// exists so that we can tell when we've gotten a zero value of it. This is used by
+	// the Call method so that we can detect when the thread is shutting down.
+	populated bool
+}
+
+type OSThread struct {
+	// call is used to submit closures for this thread to execute.
+	call chan func() (interface{}, error)
+
+	// res is used to send the results of running closures back to callers.
+	res chan execResult
+
+	// stopping is a request for this thread to stop running.
+	stopping chan struct{}
+
+	// stopped is confirmation that the thread has stopped running.
+	stopped chan struct{}
+}
+
+func NewOSThread() *OSThread {
+	return &OSThread{
+		call:     make(chan func() (interface{}, error), 1),
+		res:      make(chan execResult, 1),
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+}
+
+func (o *OSThread) start() {
+	runtime.LockOSThread()
+
+	defer func() {
+		runtime.UnlockOSThread()
+		close(o.stopped)
+	}()
+
+	for {
+		select {
+		case <-o.stopping:
+			// Close the channel used to read results so that any callers waiting on
+			// results (which can happen if they submitted something before this method
+			// has a chance to execute it but after this thread pool has been stopped)
+			// so that they immediately get a zero value which they can check for.
+			close(o.res)
```

Comment:
Why can't we just remove the buffer from `o.call`?

Reply:
> If we do, the write to it in Call() will block until it's read which may not happen if the thread is in the process of stopping.

If the process is stopping then `Call()` will go through the `case <-o.stopping` case.

If the `Call()` function has managed to push a function through `o.call` then this goroutine will always push a result back through `o.res`.

Reply:
> If the Call() function has managed to push a function through o.call then this goroutine will always push a result back through o.res.

Yes, but the write to `o.call` will block in `OSThread.Call`.

Reply:
Turns out that the test that was hanging (because the write to `o.call` was blocking was invalid. I will update the test to match how `OSThread` is supposed to be used.